### PR TITLE
Introduce OptionalRef

### PR DESCRIPTION
**Pull request by younesr1**
Introduce a OptionalRef which is intended to be similar to std::optional\<std::reference_wrapper> without the necessary layer of indirection when accessing values

### CODE REVIEW

**Comment by younesr1** [comment]
Still a WIP but reviews welcome :)

File: test/optional_ref_test.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+#include <gtest/gtest.h>
+
+namespace fixed_containers
+{
+
+/*
+*/
+TEST(OptionalRef, FixedContainerProperties)
+{
+    using T = OptionalRef<int&>;
```
Comment (alexkaratarakis) [inline test/optional_ref_test.cpp:12]: Discussion:
Should the reference be explicit as in `OptionalRef<int&>` or should it get auto-applied as in `OptionalRef<int>`

File: test/optional_ref_test.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#include "fixed_containers/optional_ref.hpp"
+
+#include <gtest/gtest.h>
+
+namespace fixed_containers
+{
+
```
Comment (alexkaratarakis) [inline test/optional_ref_test.cpp:7]: Add a test that the size is exactly 1 pointer.

File: include/fixed_containers/optional_ref.hpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+#include "fixed_containers/optional_ref_checking.hpp"
+#include "fixed_containers/concepts.hpp"
+
+#include <functional>
+#include <optional>
+
+namespace fixed_containers
+{
+    // allows T& optionals
```
Comment (alexkaratarakis) [inline include/fixed_containers/optional_ref.hpp:11]: Needs formatting.

File: include/fixed_containers/optional_ref.hpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    template <IsReference T,
+             customize::OptionalRefChecking<T> CheckingType = customize::OptionalRefAbortChecking<T>>
+    class OptionalRef
+    {
+        private:
+            using ref  = T; // T is required to be a reference type
+            using ptr = std::remove_reference<T>*;
+            using BackingType = std::optional<std::reference_wrapper<std::remove_reference_t<T>>>;
+            using Self = OptionalRef<T, CheckingType>;
+            BackingType underlying_type_; // younes todo makes this public so structural type?
```
Comment (alexkaratarakis) [inline include/fixed_containers/optional_ref.hpp:21]: I think yes, note also that a barebones `StructuralTypeOptional` already exists in https://github.com/teslamotors/fixed-containers/blob/main/include/fixed_containers/enum_utils.hpp#L259C7-L259C29 , so an `Optional` type (not just references) might be useful in the future.

File: test/optional_ref_test.cpp (hunk 1 above)
Comment (younesr1) [reply test/optional_ref_test.cpp:12]: would one point in favor of `OptionalRef<int>` be that CTAD would work correctly? we could do something like `OptionalRef maybe_foo(my_int)`, where as with the explicit one, this experssion would need to be `OptionalRef maybe_foo(std::ref(my_int))`. is this right?

Comment (zanneth) [reply test/optional_ref_test.cpp:12]: We should try to emulate existing STL APIs if possible. `std::reference_wrapper<T>` is probably the closest analog.

The code example in the documentation expresses the template argument in terms of the base type `T` and not `T&`.

```cpp
std::vector<std::reference_wrapper<int>> v(l.begin(), l.end());
```

Comment (alexkaratarakis) [reply test/optional_ref_test.cpp:12]: Familiarity with `std::reference_wrapper<T>` is a pretty good reason to do `OptionalRef<int>`.
Additionally, I tried using it with the `OptionalRef<int&>` form and indeed CTAD is problematic, and `const int&` vs `int&` is awkward as well. Also, there are paper-cuts when trying to use a `T` with implied ref, the existing comment of "using ref  = T; // T is required to be a reference type" attests to that as well.

I vote for `OptionalRef<int>` too.

Comment (younesr1) [reply test/optional_ref_test.cpp:12]: alright, will go ahead with `OptionalRef<int>`

File: include/fixed_containers/optional_ref.hpp [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+ * but without the extra indirection when dereferencing.
+ */
+template <IsNotReference T,
+          customize::OptionalRefChecking<T> CheckingType = customize::OptionalRefAbortChecking<T>>
+class OptionalRef
+{
+private:
+    using Self = OptionalRef<T, CheckingType>;
+    using ref = T&;
+    using ptr = T*;
```
Comment (zanneth) [inline include/fixed_containers/optional_ref.hpp:23]: Is there a reason why we're using the pointer type as the backing type and not simply the reference type?

File: include/fixed_containers/optional_ref.hpp [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+          customize::OptionalRefChecking<U> CheckU>
+constexpr auto operator<=>(const OptionalRef<T, CheckT>& lhs, const OptionalRef<U, CheckU>& rhs)
+{
+    if (lhs.has_value() && rhs.has_value()) return *lhs <=> *rhs;
+    if (!lhs.has_value() && !rhs.has_value()) return std::strong_ordering::equal;
+    if (rhs.has_value()) return std::strong_ordering::less;     // implies lhs doesnt have a value
+    if (lhs.has_value()) return std::strong_ordering::greater;  // implies rhs doesnt have a value
+}
+
+// <=> only gives defintion for secondary relational operators (ie. <)
```
Comment (zanneth) [inline include/fixed_containers/optional_ref.hpp:122]: `defintion` -> `definition`

File: include/fixed_containers/optional_ref.hpp [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+
+// Compares OptionalRef with a nullopt. Equivalent to above when comparing to an optional that does
+// not contain a value.
+template <typename T, customize::OptionalRefChecking<T> CheckT>
+constexpr auto operator<=>(const OptionalRef<T, CheckT>& lhs, std::nullopt_t)
+{
+    return lhs <=> OptionalRef<T, CheckT>{};
+}
+
+// <=> only gives defintion for secondary relational operators (ie. <)
```
Comment (zanneth) [inline include/fixed_containers/optional_ref.hpp:141]: `defintion` -> `definition`

File: include/fixed_containers/optional_ref.hpp [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+// Compares opt with a value. The values are compared (using the corresponding operator of T) only
+// if opt contains a value. Otherwise, opt is considered less than value. If the corresponding
+// two-way comparison expression between *opt and value is not well-formed, or if its result is not
+// convertible to bool, the program is ill-formed.
+template <typename T, customize::OptionalRefChecking<T> CheckT, class U>
+constexpr std::compare_three_way_result_t<T, U> operator<=>(const OptionalRef<T, CheckT>& lhs,
+                                                            const U& rhs)
+{
+    // using a concept gives a recursive defintion error
```
Comment (zanneth) [inline include/fixed_containers/optional_ref.hpp:156]: `defintion` -> `definition`

File: include/fixed_containers/optional_ref.hpp [hunk 9]
```diff
@@ -0,0 +1,168 @@
...
+    static_assert(std::three_way_comparable_with<T, U>);
+    return lhs.has_value() ? *lhs <=> rhs : std::strong_ordering::less;
+}
+
+// <=> only gives defintion for secondary relational operators (ie. <)
```
Comment (zanneth) [inline include/fixed_containers/optional_ref.hpp:161]: `defintion` -> `definition`

**Review by zanneth — APPROVED** [review]
LGTM except for minor typoes.

File: include/fixed_containers/optional_ref.hpp [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+    {
+    }
+
+    constexpr ~OptionalRef() = default;
+
+    constexpr OptionalRef(const Self&) noexcept = default;
+    constexpr Self& operator=(const Self&) = delete;  // copy assigment is ambiguous for this type
+
+    constexpr OptionalRef(Self&&) noexcept = default;
+    constexpr Self& operator=(Self&&) = default;
```
Comment (younesr1) [inline include/fixed_containers/optional_ref.hpp:46]: thoughts about the move assignment operator being ambiguous for this type?
ie. 
```cpp
int a = 5
OptionalRef opt(a);
int b = 10;
OptionalRef opt2(b);
opt2 = std::move(opt)
```

currently b stays at 10, and opt2 just rebinds. IMO this makes intuitive sense

Comment (alexkaratarakis) [reply include/fixed_containers/optional_ref.hpp:46]: Some relevant info;
- https://www.boost.org/doc/libs/1_85_0/libs/optional/doc/html/boost_optional/tutorial/optional_references.html
- https://stackoverflow.com/questions/26858034/why-does-stdoptional-not-have-a-specialization-for-reference-types

Comment (younesr1) [reply include/fixed_containers/optional_ref.hpp:46]: sync'd offline: We will allow assignment operators for `std::nullopt_t` and `OptionalRef<T>` but not `T`. This is motivated from this proposal
https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3406#rationale.refs

```cpp
int i = 0;
int j = 1;
optional<int&> ori;
optional<int&> orj = j;

*orj = 2;
assert (j == 2);

ori = i; // ERROR: no assignment from int&
ori = {i}; // OK: assignemnt from optional<int&>

orj = ori; // OK: rebinding assignemnt from optional<int&>
*orj = 4;
assert (j == 2);
assert (i == 4);
```

**Pull request merged**